Linking to dependencies

I'm running into compilation issues, which HPC- IT staff tells me is because "a lot of external dependencies are installed in in `lib/`, but these libraries are not used during linking. In particular, this is one of the linking lines:

````
g++ -O3 -std=c++11 -fopenmp -g -o bin/example cpp/example.pb.o 
obj/main.o -I./ -Isrc -Iobj -Icpp -L./ -L/usr/local/lib/ -lprotobuf -lz 
&& cp src/*.h* include/
```

The external protobuf library compiled in `$(TOPGIT)/lib` is not used, and 
instead the one installed in `/usr/local/lib` is."

I hope this text makes sense, and I hope there is a possible fix?

Thanks for your replies. I have tried to remove all -L flags from deps/xg/stream, but it did not help. I have to manually include a correct -Lpath flag to make things compile. 

Moreover, I seem to have trouble linking with static: cannot find -lpthread. I have removed the -static option from deps/xg/Makefile and the top level Makefile (leaving only "-static-libstdc++ -static-libgcc"). Is -static really needed?

--- a/Makefile
+++ b/Makefile
@@ -17,7 +17,7 @@ CWD:=$(shell pwd)
 
 
 LD_INCLUDE_FLAGS:=-I$(CWD)/$(INC_DIR) -I. -I$(CWD)/$(SRC_DIR) -I$(CWD)/$(UNITTEST_SRC_DIR) -I$(CWD)/$(CPP_DIR) -I$(CWD)/$(INC_DIR)/dynamic -I$(CWD)/$(INC_DIR)/sonLib
-LD_LIB_FLAGS:= -ggdb -L$(CWD)/$(LIB_DIR) -lvcflib -lgssw -lssw -lprotobuf -lhts -lpthread -ljansson -lncurses -lrocksdb -lsnappy -lz -lbz2 -lgcsa2 -lxg -ldivsufsort -ldivsufsort64 -lvcfh -lgfakluge -lraptor2 -lsupbub -lsdsl -lpinchesandcacti -l3edgeconnected  -lsonlib
+LD_LIB_FLAGS:= -ggdb -L$(CWD)/$(LIB_DIR) -Wl,-rpath,$(CWD)/$(LIB_DIR)  -lvcflib -lgssw -lssw -lprotobuf -lhts -lpthread -ljansson -lncurses -lrocksdb -lsnappy -lz -lbz2 -lgcsa2 -lxg -ldivsufsort -ldivsufsort64 -lvcfh -lgfakluge -lraptor2 -lsupbub -lsdsl -lpinchesandcacti -l3edgeconnected  -lsonlib
 
 ifeq ($(shell uname -s),Darwin)
     # We may need libraries from Macports

--- a/Makefile
+++ b/Makefile
@@ -17,7 +17,7 @@ CWD:=$(shell pwd)
 
 
 LD_INCLUDE_FLAGS:=-I$(CWD)/$(INC_DIR) -I. -I$(CWD)/$(SRC_DIR) -I$(CWD)/$(UNITTEST_SRC_DIR) -I$(CWD)/$(CPP_DIR) -I$(CWD)/$(INC_DIR)/dynamic -I$(CWD)/$(INC_DIR)/sonLib
-LD_LIB_FLAGS:= -ggdb -L$(CWD)/$(LIB_DIR) -Wl,-rpath,$(CWD)/$(LIB_DIR)  -lvcflib -lgssw -lssw -lprotobuf -lhts -lpthread -ljansson -lncurses -lrocksdb -lsnappy -lz -lbz2 -lgcsa2 -lxg -ldivsufsort -ldivsufsort64 -lvcfh -lgfakluge -lraptor2 -lsupbub -lsdsl -lpinchesandcacti -l3edgeconnected  -lsonlib
+LD_LIB_FLAGS:= -ggdb -L$(CWD)/$(LIB_DIR) -lvcflib -lgssw -lssw -lprotobuf -lhts -lpthread -ljansson -lncurses -lrocksdb -lsnappy -lz -lbz2 -lgcsa2 -lxg -ldivsufsort -ldivsufsort64 -lvcfh -lgfakluge -lraptor2 -lsupbub -lsdsl -lpinchesandcacti -l3edgeconnected  -lsonlib
 
 ifeq ($(shell uname -s),Darwin)
     # We may need libraries from Macports
@@ -29,9 +29,12 @@ ifeq ($(shell uname -s),Darwin)
        LD_LIB_FLAGS += -L/usr/local/lib
     endif
     ROCKSDB_PORTABLE=PORTABLE=1 # needed to build rocksdb without weird assembler options
+    # TODO: configure RPATH-equivalent on OS X for finding libraries without environment variables at runtime
 else
     # Not on OS X, we can have librt
     LD_LIB_FLAGS += -lrt
+    # We can also have a normal Unix rpath
+    LD_LIB_FLAGS += -Wl,-rpath,$(CWD)/$(LIB_DIR)  
 endif
 
 STATIC_FLAGS=-static -static-libstdc++ -static-libgcc

